### PR TITLE
Implement `bal run --watch`.

### PR DESCRIPTION
**Pull request by gayaldassanayake**
## Purpose
$ subject

Related to https://github.com/ballerina-platform/ballerina-lang/issues/42892

## Approach

image

## Samples

https://github.com/user-attachments/assets/ced3ffc4-dcc8-4296-815f-52f1c94b1ae9

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ProjectWatcher.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+                Path changedFilePath = dir.resolve(changedFileName);
+                if (isValidFileChange(changedFilePath)) {
+                    thread.terminate();
+                    thread.join();
+                    thread = new RunCommandExecutor(runCommand);
+                    thread.start();
+                }
+                if (kind == ENTRY_CREATE) {
+                    if (Files.isDirectory(changedFilePath)) {
+                        registerFileTree(changedFilePath);
```
Comment (Thevakumar-Luheerathan) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ProjectWatcher.java:82]:
```suggestion
                if (kind == ENTRY_CREATE) && Files.isDirectory(changedFilePath) {
                        registerFileTree(changedFilePath);
```

Comment (gayaldassanayake) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ProjectWatcher.java]: Fixed with 79836775a84e991a397826174db265f781abd64e

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ProjectWatcher.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    private boolean isValidFileChange(Path path) {
+        if (Files.isDirectory(path)) {
+            return false;
+        }
+        Path fileNamePath = path.getFileName();
+        if (fileNamePath == null) {
+            return false;
+        }
+        String fileName = fileNamePath.toString();
+        return fileName.endsWith(".bal") || fileName.endsWith(".toml") && !fileName.equals("Dependencies.toml");
```
Comment (Thevakumar-Luheerathan) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ProjectWatcher.java:103]: can introduce constants

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/RunCommand.java [hunk 3]
```diff
@@ -83,6 +84,8 @@ public class RunCommand implements BLauncherCmd {
     @CommandLine.Option(names = "--debug", hidden = true)
     private String debugPort;
 
+    @CommandLine.Option(names = "--watch", description = "listen for file changes and automatically re-run the project")
```
Comment (Thevakumar-Luheerathan) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/RunCommand.java:87]:
```suggestion
    @CommandLine.Option(names = "--watch", description = "listen for file changes and automatically rerun the project")
```

**Comment by gayaldassanayake** [comment]
TODOs
- [ ] Write unit tests
- [ ] UX of the command

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ProjectWatcher.java (hunk 2 above)
Comment (SasinduDilshara) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ProjectWatcher.java:103]: We can omit the tests file changes. If the change in test directory we can omit that. wdyt?

Comment (SasinduDilshara) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ProjectWatcher.java:103]: If a file in the resource directory got chanaged, don't we need to consider that? wdyt?

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/RunCommand.java (hunk 3 above)
Comment (gayaldassanayake) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/RunCommand.java]: Changed listen to watch. According to chatgpt,

> The correct term is "re-run" when used as a verb, meaning to run again. So, the phrase should be:
> 
> "watch for file changes and automatically re-run the project"

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/RunCommand.java [hunk 4]
```diff
@@ -319,4 +346,11 @@ private BuildOptions constructBuildOptions() {
 
         return buildOptionsBuilder.build();
     }
+
+    public boolean containsService() {
+        if (project != null) {
+            return ProjectUtils.containsDefaultModuleService(project.currentPackage());
+        }
+        return true;
```
Comment (SasinduDilshara) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/RunCommand.java:354]: return project == null || ProjectUtils.containsDefaultModuleService(project.currentPackage());

Just a suggestion

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/watch/ProjectWatcher.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
+/*
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/watch/ProjectWatcher.java:1]: Let's move this and RunCommandExecutor classes to `io.ballerina.cli.utils` package

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/watch/ProjectWatcher.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+            for (WatchEvent<?> event : key.pollEvents()) {
+                WatchEvent.Kind<?> kind = event.kind();
+                if (kind == OVERFLOW) {
+                    continue;
+                }
+                WatchEvent<Path> pathWatchEvent = cast(event);
+                Path changedFileName = pathWatchEvent.context();
+                Path changedFilePath = dir.resolve(changedFileName).toAbsolutePath();
+                if (isValidFileChange(changedFilePath)) {
+                    outStream.println("\nDetected file changes. Re-running the project.");
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/watch/ProjectWatcher.java:92]:
```suggestion
                    outStream.println("\nDetected file changes. Re-running the project...");
```

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/RunCommand.java (hunk 4 above)
Comment (gayaldassanayake) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/RunCommand.java]: Fixed with 9305aa7069635a900c85be9dcfd5d7a0452b032d

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/watch/ProjectWatcher.java (hunk 5 above)
Comment (gayaldassanayake) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/watch/ProjectWatcher.java]: Fixed with 9305aa7069635a900c85be9dcfd5d7a0452b032d

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/watch/ProjectWatcher.java (hunk 6 above)
Comment (gayaldassanayake) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/watch/ProjectWatcher.java]: Fixed with 9305aa7069635a900c85be9dcfd5d7a0452b032d

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/ProjectWatcher.java [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+            for (WatchEvent<?> event : key.pollEvents()) {
+                WatchEvent.Kind<?> kind = event.kind();
+                if (kind == OVERFLOW) {
+                    continue;
+                }
+                WatchEvent<Path> pathWatchEvent = cast(event);
+                Path changedFileName = pathWatchEvent.context();
+                Path changedFilePath = dir.resolve(changedFileName).toAbsolutePath();
+                if (isValidFileChange(changedFilePath)) {
+                    outStream.println("\\nDetected file changes. Re-running the project...");
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/ProjectWatcher.java:92]: This prints `\nDetected file changes. Re-running the project...`. Need to remove the leading \n

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/ProjectWatcher.java [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+import static java.nio.file.StandardWatchEventKinds.OVERFLOW;
+
+/**
+ * Represents a watcher that looks into project file changes.
+ *
+ * @since 2201.10.0
+ */
+public class ProjectWatcher {
+    private static final String BAL_EXTENSION = ".bal";
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/ProjectWatcher.java:50]: There is a constant defined in the ProjectConstants. We can use that.

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/ProjectWatcher.java [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+    private static final String TOML_EXTENSION = ".toml";
+    private static final String DEPENDENCIES_TOML = "Dependencies.toml";
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/ProjectWatcher.java:52]: There is a constant defined in the ProjectConstants. We can use that.

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/ProjectWatcher.java [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+import static java.nio.file.StandardWatchEventKinds.OVERFLOW;
+
+/**
+ * Represents a watcher that looks into project file changes.
+ *
+ * @since 2201.10.0
+ */
+public class ProjectWatcher {
+    private static final String BAL_EXTENSION = ".bal";
+    private static final String TOML_EXTENSION = ".toml";
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/ProjectWatcher.java:51]: Let's move this to ProjectConstants class

Comment (gayaldassanayake) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/ProjectWatcher.java]: Resolved with 3d007562f6199380baff4b3b43bb29cbd1b7709c

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/ProjectWatcher.java (hunk 9 above)
Comment (gayaldassanayake) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/ProjectWatcher.java]: Resolved with 3d007562f6199380baff4b3b43bb29cbd1b7709c

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/ProjectWatcher.java (hunk 8 above)
Comment (gayaldassanayake) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/ProjectWatcher.java]: Resolved with 3d007562f6199380baff4b3b43bb29cbd1b7709c

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/ProjectWatcher.java (hunk 7 above)
Comment (gayaldassanayake) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/utils/ProjectWatcher.java]: Resolved with 3d007562f6199380baff4b3b43bb29cbd1b7709c

**Comment by gayaldassanayake** [comment]
I will introduce the test cases in a different PR. Due to the multithreaded nature, test cases should be written in a thread-safe manner.

**Review by azinneera — APPROVED** [review]

**Review by Dilhasha — APPROVED** [review]

**Pull request merged**